walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,102 @@
 import java.util.Collection;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.List;
 
 public final class FindMeetingQuery {
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+  	/** General algorithm (first attempt, could be better optimized, generally runs in linear
+     * time except for calls to Collections.disjoint() in the getAllRelevantEvents() method): 
+     * 1. Get a list of all of the events which actually concern the attendees of the request.
+     * 2. Sort the list by start time.
+     * 3. Loop through, and find all available times "between" these events; requires some amount
+     * of shifting around bounds to deal with edge cases regarding nested events.
+     * 4. Return a list of these times, only adding a time if it is longer than the proposed
+     * duration of the MeetingRequest. 
+     */
+  	Collection<Event> relevantEvents = getAllRelevantEvents(events, request);
+    List<TimeRange> busyTimes = getSortedTimeRanges(relevantEvents);
+    return getComplementTimes(busyTimes, request);
   }
+
+  /**
+   * Returns a Collection of only those events which have
+   * attendees specified in the request parameter.
+   */
+  private Collection<Event> getAllRelevantEvents(Collection<Event> events, MeetingRequest request) {
+    Collection<String> attendees = request.getAttendees();
+    Collection<Event> relevantEvents = new HashSet<>();
+    for (Event event : events) {
+      Collection<String> eventAttendees = event.getAttendees();
+      if (!Collections.disjoint(attendees, eventAttendees)) {
+        relevantEvents.add(event);
+      }
+    }
+    return relevantEvents;
+  }
+
+  /**
+  * Returns a sorted (by start time) List<TimeRange> representing the Collection of events
+  * passed as a parameter. 
+  */
+  private List<TimeRange> getSortedTimeRanges(Collection<Event> events) {
+    List<TimeRange> times = new ArrayList<>();
+    for (Event event : events) {
+      times.add(event.getWhen());
+    }
+    Collections.sort(times, TimeRange.ORDER_BY_START);
+    return times;
+  }
+
+  /** Returns a Collection of only those times which do not overlap with any TimeRange in times */
+	private Collection<TimeRange> getComplementTimes(List<TimeRange> times, MeetingRequest request) {
+    Collection<TimeRange> complementTimes = new ArrayList<>();
+    if (times.size() == 0) {
+      if (proposedTimeIsLongEnough(TimeRange.WHOLE_DAY, request)) {
+        complementTimes.add(TimeRange.WHOLE_DAY);
+      }
+      return complementTimes;
+	  }
+    // Add any time before the first event in the list
+    if (! (TimeRange.START_OF_DAY == times.get(0).start())) {
+      TimeRange proposedSlot = TimeRange.fromStartEnd(TimeRange.START_OF_DAY, times.get(0).start(), false);
+        if (proposedTimeIsLongEnough(proposedSlot, request)) {
+          complementTimes.add(proposedSlot);
```

Comment:
You seem to be repeating this paradigm a lot (i.e. create TimeRange, if proposedTimeIsLongEnough, add).  It might be better to combine this all into proposedTimeIsLongEnough or a bigger helper function.